File: ai-service/app/agents/supervisor_agent.py
from __future__ import annotations

import json
import time
from typing import Any

from app.agents.specialist_agent import run_specialist_agent
from app.repositories import agent_repository, learning_repository
from app.tools.base import ToolContext
from app.tools.performance import GetStudentPerformanceTool

_perf_tool = GetStudentPerformanceTool()


def run_supervisor_agent(
    supervisor_run_id: str,
    student_id: str,
    goal: str,
    triggered_by_user_id: str | None,
    spec_def: dict[str, Any],
) -> dict[str, Any]:

    # ── Step 1 (DECISION): START ──────────────────────────────────────────────
    agent_repository.upsert_completed_step(
        supervisor_run_id, 1, "DECISION", None,
        json.dumps({"goal": goal}),
        json.dumps({"decision": "START"}),
        "COMPLETED", None, None, 0,
    )

    # ── Step 2 (TOOL): GetStudentPerformance — direct call, no LLM ───────────
    agent_repository.insert_running_tool_step(
        supervisor_run_id, 2, "GetStudentPerformance",
        json.dumps({"studentId": student_id}),
    )
    perf_ctx = ToolContext(student_id=student_id, agent_run_id=supervisor_run_id)
    perf_start = time.time()
    perf_result = _perf_tool.execute({"studentId": student_id}, perf_ctx)
    perf_ms = int((time.time() - perf_start) * 1000)

    agent_repository.update_tool_step(
        supervisor_run_id, 2,
        json.dumps(perf_result.data) if perf_result.success else None,
        "COMPLETED" if perf_result.success else "FAILED",
        perf_result.error_code if not perf_result.success else None,
        perf_result.error_message if not perf_result.success else None,
        perf_ms,
    )

    # ── Step 3 (DECISION): DELEGATE_TO_SPECIALIST ─────────────────────────────
    agent_repository.upsert_completed_step(
        supervisor_run_id, 3, "DECISION", None,
        json.dumps({"decision": "DELEGATE_TO_SPECIALIST"}),
        json.dumps({"reason": "learning plan requires specialist analysis"}),
        "COMPLETED", None, None, 0,
    )

    specialist_result = run_specialist_agent(
        {
            "studentId": student_id,
            "goal": goal,
            "supervisorRunId": supervisor_run_id,
            "triggeredByUserId": triggered_by_user_id,
        },
        spec_def,
    )

    if not specialist_result.get("draft_plan"):
        raise RuntimeError("Specialist failed to produce a learning plan")

    # ── Step 4 (DECISION): PERSIST_PLAN ──────────────────────────────────────
    agent_repository.upsert_completed_step(
        supervisor_run_id, 4, "DECISION", None,
        json.dumps({"decision": "PERSIST_PLAN"}),
        json.dumps({"specialistRunId": specialist_result["specialist_run_id"]}),
        "COMPLETED", None, None, 0,
    )

    # ── Idempotency guard ─────────────────────────────────────────────────────
    existing_plan = learning_repository.find_plan_by_agent_run(supervisor_run_id)

    if existing_plan:
        learning_plan = existing_plan
    else:
        learning_plan = learning_repository.create_learning_plan(
            student_id=student_id,
            agent_run_id=supervisor_run_id,
            goal=goal,
            plan_data_json_str=json.dumps(specialist_result["draft_plan"]),
        )
        for ws in specialist_result.get("weak_skills") or []:
            skill_id = ws.get("skill_id") or str(ws.get("skill_id", ""))
            name = ws.get("name", "Unknown skill")
            learning_repository.create_recommendation(
                student_id=student_id,
                plan_id=learning_plan["id"],
                skill_id=skill_id,
                title=f"Improve: {name}",
            )

    return {
        "learning_plan": learning_plan,
        "specialist_result": specialist_result,
    }

File: backend/src/agents/supervisorAgent.ts
import { db } from '../shared/db/pool';
import { getStudentPerformanceTool } from './tools';
import { runSpecialistAgent, SpecialistResult } from './specialistAgent';

// ── Public types ──────────────────────────────────────────────────────────────

export interface SupervisorInput {
  supervisorRunId: string;
  studentId: string;
  goal: string;
  triggeredByUserId: string | null;
  specDef: {
    id: string;
    max_steps: number;
    max_tool_calls: number;
    timeout_seconds: number;
  };
}

export interface SupervisorResult {
  learningPlan: Record<string, unknown>;
  specialistResult: SpecialistResult;
}

// ── Supervisor agent ──────────────────────────────────────────────────────────

export async function runSupervisorAgent(
  input: SupervisorInput
): Promise<SupervisorResult> {
  const { supervisorRunId, studentId, goal, triggeredByUserId, specDef } = input;

  // ── Step 1 (DECISION): Supervisor starts, will gather basic context ───────
  await insertStep(
    supervisorRunId, 1, 'DECISION', null,
    { goal },
    { decision: 'START' },
    'COMPLETED', null, null, 0
  );

  // ── Step 2 (TOOL): GetStudentPerformance — direct call, no LLM loop ───────
  // Supervisor has read access to performance data for routing context only.
  await db.query(
    `INSERT INTO agent.agent_steps
       (agent_run_id, sequence_no, step_type, tool_name, input, status)
     VALUES ($1,2,'TOOL','GetStudentPerformance',$2,'RUNNING')`,
    [supervisorRunId, JSON.stringify({ studentId })]
  );

  const perfStart = Date.now();
  const perfCtx   = { studentId, agentRunId: supervisorRunId };
  const perfResult = await getStudentPerformanceTool.execute({ studentId }, perfCtx);
  const perfMs    = Date.now() - perfStart;

  await db.query(
    `UPDATE agent.agent_steps
     SET output=$2, status=$3, duration_ms=$4
     WHERE agent_run_id=$1 AND sequence_no=2`,
    [
      supervisorRunId,
      JSON.stringify(perfResult.data ?? null),
      perfResult.success ? 'COMPLETED' : 'FAILED',
      perfMs,
    ]
  );

  // ── Step 3 (DECISION): Delegate to specialist ─────────────────────────────
  await insertStep(
    supervisorRunId, 3, 'DECISION', null,
    { decision: 'DELEGATE_TO_SPECIALIST' },
    { reason: 'learning plan requires specialist analysis' },
    'COMPLETED', null, null, 0
  );

  const specialist = await runSpecialistAgent(
    { studentId, goal, supervisorRunId, triggeredByUserId },
    specDef
  );

  if (!specialist.draftPlan) {
    throw new Error('Specialist failed to produce a learning plan');
  }

  // ── Step 4 (DECISION): Validate and persist plan ──────────────────────────
  await insertStep(
    supervisorRunId, 4, 'DECISION', null,
    { decision: 'PERSIST_PLAN' },
    { specialistRunId: specialist.specialistRunId },
    'COMPLETED', null, null, 0
  );

  // ── Idempotency guard: do not create a duplicate plan ────────────────────
  const { rows: existing } = await db.query(
    `SELECT id FROM performance.learning_plans WHERE generated_by_agent_run_id = $1`,
    [supervisorRunId]
  );

  let learningPlan: Record<string, unknown>;

  if (existing.length > 0) {
    learningPlan = existing[0] as Record<string, unknown>;
  } else {
    const { rows: planRows } = await db.query(
      `INSERT INTO performance.learning_plans
         (student_id, generated_by_agent_run_id, goal, plan_data, status, version)
       VALUES ($1,$2,$3,$4,'ACTIVE',1)
       RETURNING *`,
      [studentId, supervisorRunId, goal, JSON.stringify(specialist.draftPlan)]
    );
    learningPlan = planRows[0] as Record<string, unknown>;

    for (const ws of specialist.weakSkills) {
      await db.query(
        `INSERT INTO performance.learning_recommendations
           (student_id, learning_plan_id, skill_id, recommendation_type,
            title, description, priority, status)
         VALUES ($1,$2,$3,'SKILL_GAP',$4,
                 'Focus on this skill based on performance gap',1,'ACTIVE')`,
        [studentId, learningPlan.id, ws.skill_id, `Improve: ${ws.name}`]
      );
    }
  }

  return { learningPlan, specialistResult: specialist };
}

// ── Local step persistence helper ─────────────────────────────────────────────

async function insertStep(
  runId: string,
  seqNo: number,
  stepType: string,
  toolName: string | null,
  input: unknown,
  output: unknown,
  status: string,
  errorCode: string | null,
  errorMessage: string | null,
  durationMs: number
): Promise<void> {
  await db.query(
    `INSERT INTO agent.agent_steps
       (agent_run_id, sequence_no, step_type, tool_name, input, output,
        status, error_code, error_message, duration_ms)
     VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10)
     ON CONFLICT (agent_run_id, sequence_no) DO NOTHING`,
    [
      runId, seqNo, stepType, toolName,
      JSON.stringify(input ?? null),
      output !== null ? JSON.stringify(output) : null,
      status, errorCode, errorMessage, durationMs,
    ]
  );
}
